tutorials/haskell/servant-auth/tutorial.md

```diff
@@ -0,0 +1,566 @@
+instance:
+
+```haskell
+type instance AuthCookieData = Maybe Session
+```
+
+Now we are ready to write the check whose type is `AuthHandler Request
+(Maybe Session)`:
+
+```haskell
+-- | The acutal session check. The ‘servant-auth-cookie’ package provides
+-- the 'defaultAuthHandler', but that does not indicate missing 'Session' as
+-- 'Nothing', so we use the custom one.
+
+cookieAuthCheck
+  :: AuthCookieSettings
+  -> ServerKey
+  -> AuthHandler Request (Maybe Session)
+cookieAuthCheck authSettings serverKey = mkAuthHandler $ \request -> do
+  result <- try (getSession authSettings serverKey request)
+  case result :: Either AuthCookieException (Maybe Session) of
+    Left _ -> return Nothing
+    Right session -> return session
+```
+
+Now we can perform checks whether cookie is present and can be de-serialized
+or not. Then goes some boilperplate:
+
+```haskell
+
+-- | The entry point of the application.
+
+main :: IO ()
+main = do
+  let appPort = 8000
+  randomSource <- mkRandomSource drgNew 2000
+  serverKey    <- mkServerKey 16 (Just $ fromIntegral (86400 :: Integer))
+  let authSettings = AuthCookieSettings
+        { acsSessionField = "Session"
+        , acsCookieFlags  = ["HttpOnly"]
+        , acsMaxAge       = fromIntegral (6 * 3600 :: Integer)
+        , acsExpirationFormat = "%0Y%m%d%H%M%S"
+        , acsPath         = "/"
+        , acsHashAlgorithm = Proxy :: Proxy SHA256
+        , acsCipher       = Proxy :: Proxy AES256
+        , acsEncryptAlgorithm = ctrCombine
+        , acsDecryptAlgorithm = ctrCombine }
+  run appPort . app $ AppContext
+    { appContextAuthSettings = authSettings
+    , appContextRandomSource = randomSource
+    , appContextServerKey    = serverKey
+    , appContextApproot      = "localhost"
+    , appContextPort         = appPort
+    , appContextScheme       = "http:" }
+
+app :: AppContext -> Application
+app appContext@AppContext {..} = serveWithContext
+  (Proxy :: Proxy Routes)
+  ((cookieAuthCheck appContextAuthSettings appContextServerKey
+     :: AuthHandler Request (Maybe Session))
+   :. EmptyContext)
+  (server appContext)
+
+server :: AppContext -> Server Routes
+server context = enter (Nat $ flip runReaderT context) handlers
+```
+
+Note that we create `RandomSource` which is a source of random numbers that
+will reset itself after producing 2000 random bytes. The source can be used
+in concurrent code because the actual modification of its internal state is
+atomic. `ServerKey` is another thing we are going to need for cookie
+encryption, it has the ability to “expire” after specified period of time
+(one day in our case, but if we passed `Nothing` it would never expire),
+it's suitable for concurrent usage as well.
+
+Other things to note is the use of `serveWithContext` that is
+well-documented in the Servant documentation
+[here](https://www.stackage.org/haddock/nightly-2016-08-06/servant-server-0.7.1/Servant-Server.html#v:serveWithContext)
```

Comment:
@mrkkrp Is the well documented part just for the `serveWithContext` function or for the module in general? I don't see any documentation for the specific function.